golem-service-base/src/service/template_object_store.rs

```diff
@@ -78,18 +100,34 @@ impl TemplateObjectStore for AwsS3TemplateObjectStore {
         Ok(data.to_vec())
     }
 
-    async fn put(&self, object_key: &str, data: Vec<u8>) -> Result<(), Box<dyn Error>> {
+    async fn get_stream(&self, object_key: &str) -> ByteStream {
         let key = self.get_key(object_key);
 
-        info!("Putting object: {}/{}", self.bucket_name, key);
+        info!("Getting object: {}/{}", self.bucket_name, key);
+
+        match self
+            .client
+            .get_object()
+            .bucket(&self.bucket_name)
+            .key(key)
+            .send()
+            .await
+        {
+            Ok(response) => response.body.into(),
```

Comment:
How is the whole impl different to using `into_async_read` https://dtantsur.github.io/rust-openstack/tokio/io/trait.AsyncRead.html

Reply:
in case of gRpc Api returning stream, 
then we had to transform AsyncRead back to stream

Reply:
AsyncRead can be Stream. But just looking through the APIs. take it with a grain salt as I have never done this before and may be reading these APIs ina. wrong way. But ofcourse you can see if the following works, without much wrappers of our own, if you are interested (you already got approvals :) )

```
    async fn get(&self, object_key: &str) -> Result<impl AsyncRead, Box<dyn Error>>;



```

```
 let response = self
            .client
            .get_object()
            .bucket(&self.bucket_name)
            .key(key)
            .send()
            .await?;

        let body = response.body.into_async_read();
        Ok(body)
```

```
        let bytes: impl AsyncRead+Sized = self
            .template_service
            .download(&template_id.0, version.0)
            .await?;

        Ok(Response::builder()
            .header("Content-Type", "application/octet-stream")
            .body(Body::from_bytes_stream(bytes))
            .expect("failed to construct response"))
```

Reply:
but this is more like case of rest api
where this can be done 

```rust
 async fn download_template(
        &self,
        template_id: Path<TemplateId>,
        version: Query<Option<u64>>,
    ) -> Result<Binary<Body>> {
        let bytes = self
            .template_service
            .download_stream(&template_id.0, version.0)
            .await?;
        Ok(Binary(Body::from_bytes_stream(bytes.map_err(|e| {
            std::io::Error::new(std::io::ErrorKind::Other, e.to_string())
        }))))
    }
```